### PR TITLE
dont explode when sorting nil property

### PR DESCRIPTION
**Pull request by richardmonette**
Potential fix for https://github.com/Shopify/shopify/issues/84192

### CODE REVIEW

File: lib/liquid/standardfilters.rb [hunk 1]
```diff
@@ -125,7 +125,7 @@ def sort(input, property = nil)
       elsif ary.empty? # The next two cases assume a non-empty array.
         []
       elsif ary.first.respond_to?(:[]) && !ary.first[property].nil?
-        ary.sort { |a, b| a[property] <=> b[property] }
+        ary.sort { |a, b| a[property] && b[property] ? a[property] <=> b[property] : a[property] ? -1 : 1 }
```
Comment (fw42) [inline lib/liquid/standardfilters.rb]: this line is way to confusing... can we add some parentheses?

**Comment by fw42** [comment]
can we have a test to assert the behaviour of comparing nils to other nils? what should happen in that case?

File: lib/liquid/standardfilters.rb (hunk 1 above)
Comment (richardmonette) [inline lib/liquid/standardfilters.rb]: I can break it up across a few lines instead of using compound ternary statements. Basically, its just saying, if we have values for both properties, then use the combined comparison operator, otherwise, if the first element has the value, its considered larger.

**Comment by richardmonette** [comment]
@fw42 Can add a test but, what would the expectation be? In my opinion, when sorting, and some values are nil, the sortable values should come first in order, and then the nils, in the case of ascending. The nils would be considered the lowest possible value. Order amongst the nil values would not be guaranteed, as there is no sortable value.

**Comment by richardmonette** [comment]
If we are not comfortable with having the nils portion of the sorted array be non-deterministic, I would propose we extend the `ary.first.respond_to?(:[]) && !ary.first[property].nil?` check to make sure all values pass this check, otherwise we raise an error saying the input cannot be sorted on the given property.

**Comment by richardmonette** [comment]
cc @pushrax @dylanahsmith

**Comment by pushrax** [comment]
>  In my opinion, when sorting, and some values are nil, the sortable values should come first in order, and then the nils, in the case of ascending. The nils would be considered the lowest possible value.

isn't this contradictory?

File: lib/liquid/standardfilters.rb (hunk 1 above)
Comment (pushrax) [inline lib/liquid/standardfilters.rb]: It would also be nice to avoid evaluating `[]` multiple times. 👍  to breaking it up into multiple lines, and also caching the value of `[property]`.

**Comment by richardmonette** [comment]
@pushrax Sorry, my explanation might not be worded as clearly as possible. I've updated the test to try and better convey the intent of the change. 

The example is something like, say we have 3 search results, 2 ProductDrops and 1 ArticleDrop, a given sorting property of 'price'. 

It seems a reasonable result would be something like, the cheapest product first, then the second cheapest, then the ArticleDrop, where the price cannot be determined.

As it stands, this would just caused `ArgumentError: comparison of Hash with Hash failed` which doesn't seem great.

**Comment by fw42** [comment]
I'm ok with this. If @pushrax is too, then 👍 🚢

**Comment by pushrax** [comment]
:+1:
